### PR TITLE
Fix 'replace missing metadata' for music

### PR DESCRIPTION
**Pull request by GinoGinocchio**
**Changes**
The code includes logic that ensures that manually entered data by the user in certain metadata fields is not modified during the process of searching and updating missing metadata

**Issues**
Fixes issue #9760 for music.

### CODE REVIEW

File: MediaBrowser.Providers/MediaInfo/AudioFileProber.cs [hunk 1]
```diff
@@ -250,14 +250,19 @@ private void FetchDataFromTags(Audio audio)
...
+                    audio.Artists = audio.Artists != null ? audio.Artists : performers;
+                    // audio.AlbumArtists = albumArtists;
+                    audio.AlbumArtists = audio.AlbumArtists != null ? audio.AlbumArtists : albumArtists;
                 }
 
-                audio.Name = tags.Title;
-                audio.Album = tags.Album;
-                audio.IndexNumber = Convert.ToInt32(tags.Track);
-                audio.ParentIndexNumber = Convert.ToInt32(tags.Disc);
+                audio.Name = string.IsNullOrEmpty(audio.Name.ToString()) ? tags.Title : audio.Name;
```
Comment (cvium) [inline MediaBrowser.Providers/MediaInfo/AudioFileProber.cs:259]: Why ToString?

File: MediaBrowser.Providers/MediaInfo/AudioFileProber.cs [hunk 2]
```diff
@@ -250,14 +250,19 @@ private void FetchDataFromTags(Audio audio)
                     }
 
                     _libraryManager.UpdatePeople(audio, people);
-                    audio.Artists = performers;
-                    audio.AlbumArtists = albumArtists;
+                    // audio.Artists = performers;
```
Comment (cvium) [inline MediaBrowser.Providers/MediaInfo/AudioFileProber.cs:253]: Remove commented out code

File: MediaBrowser.Providers/MediaInfo/AudioFileProber.cs [hunk 3]
```diff
@@ -271,11 +276,30 @@ private void FetchDataFromTags(Audio audio)
                     audio.Genres = tags.Genres.Distinct(StringComparer.OrdinalIgnoreCase).ToArray();
                 }
 
-                audio.SetProviderId(MetadataProvider.MusicBrainzArtist, tags.MusicBrainzArtistId);
-                audio.SetProviderId(MetadataProvider.MusicBrainzAlbumArtist, tags.MusicBrainzReleaseArtistId);
-                audio.SetProviderId(MetadataProvider.MusicBrainzAlbum, tags.MusicBrainzReleaseId);
-                audio.SetProviderId(MetadataProvider.MusicBrainzReleaseGroup, tags.MusicBrainzReleaseGroupId);
-                audio.SetProviderId(MetadataProvider.MusicBrainzTrack, tags.MusicBrainzTrackId);
+                if (!audio.ProviderIds.TryGetValue("MusicBrainzArtist", out var artistValue) || artistValue == null)
```
Comment (cvium) [inline MediaBrowser.Providers/MediaInfo/AudioFileProber.cs:279]: Use TryGetProviderId

File: MediaBrowser.Providers/MediaInfo/AudioFileProber.cs [hunk 4]
```diff
@@ -271,27 +271,27 @@ private void FetchDataFromTags(Audio audio)
                     audio.Genres = tags.Genres.Distinct(StringComparer.OrdinalIgnoreCase).ToArray();
                 }
 
-                if (!audio.ProviderIds.TryGetValue("MusicBrainzArtist", out var artistValue) || artistValue == null)
+                if (!audio.TryGetProviderId("MusicBrainzArtist", out var artistValue) || string.IsNullOrEmpty(artistValue))
```
Comment (cvium) [inline MediaBrowser.Providers/MediaInfo/AudioFileProber.cs:274]: You do not need to check for null nor emptiness. If TryGetProviderId returns true, then it is a valid provider id.

File: MediaBrowser.Providers/MediaInfo/AudioFileProber.cs [hunk 5]
```diff
@@ -250,14 +250,14 @@ private void FetchDataFromTags(Audio audio)
                     }
 
                     _libraryManager.UpdatePeople(audio, people);
-                    audio.Artists = performers;
-                    audio.AlbumArtists = albumArtists;
+                    audio.Artists = audio.Artists != null ? audio.Artists : performers;
+                    audio.AlbumArtists = audio.AlbumArtists != null ? audio.AlbumArtists : albumArtists;
```
Comment (crobibero) [inline MediaBrowser.Providers/MediaInfo/AudioFileProber.cs:254]:
```suggestion
                    audio.Artists ??= performers;
                    audio.AlbumArtists ??= albumArtists;
```

File: MediaBrowser.Providers/MediaInfo/AudioFileProber.cs [hunk 6]
```diff
@@ -250,14 +250,14 @@ private void FetchDataFromTags(Audio audio)
...
                 }
 
-                audio.Name = tags.Title;
-                audio.Album = tags.Album;
-                audio.IndexNumber = Convert.ToInt32(tags.Track);
-                audio.ParentIndexNumber = Convert.ToInt32(tags.Disc);
+                audio.Name = string.IsNullOrEmpty(audio.Name) ? tags.Title : audio.Name;
+                audio.Album = audio.Album != null ? audio.Album : tags.Album;
+                audio.IndexNumber = audio.IndexNumber != null ? audio.IndexNumber : Convert.ToInt32(tags.Track);
+                audio.ParentIndexNumber = audio.ParentIndexNumber != null ? audio.ParentIndexNumber : Convert.ToInt32(tags.Disc);
```
Comment (crobibero) [inline MediaBrowser.Providers/MediaInfo/AudioFileProber.cs:260]:
```suggestion
                audio.Album ??= tags.Album;
                audio.IndexNumber ??= Convert.ToInt32(tags.Track);
                audio.ParentIndexNumber ??= Convert.ToInt32(tags.Disc);
```

**Review by IDisposable — COMMENTED** [review]
THANK YOU! This is an awesome fix :)

**Review by crobibero — APPROVED** [review]

**Review by Shadowghost — APPROVED** [review]

**Review by Bond-009 — APPROVED** [review]

**Review by lonebyte — COMMENTED** [review]
Won't this also prevent updating if the metadata should actually be replaced? I think `MetadataRefreshOptions.ReplaceAllMetadata` needs to be evaluated somewhere.

**Comment by cvium** [comment]
> Won't this also prevent updating if the metadata should actually be replaced? I think `MetadataRefreshOptions.ReplaceAllMetadata` needs to be evaluated somewhere.

I think this is an excellent point

**Review by cvium — CHANGES_REQUESTED** [review]
See last comment

**Comment by Shadowghost** [comment]
@GinoGinocchio are you able to look further into this?

**Comment by crobibero** [comment]
Closing in favor of https://github.com/jellyfin/jellyfin/pull/10691

**Pull request closed without merging**